File: LLMgeneratedCode/gemini-2.R
# Load necessary libraries
library(dplyr)
library(glmnet)
library(pROC)

# --- PTB Prediction ---

# 1. Data Loading and Preprocessing for PTB

# Load data
train_features_ptb_raw <- read.csv("training_species_abundance.csv")
validation_features_ptb_raw <- read.csv("validation_species_abundance.csv")
train_metadata_ptb <- read.csv("training_metadata.csv")
validation_metadata_ptb <- read.csv("validation_metadata.csv")

# Merge features and metadata for training set
train_data_ptb_all <- merge(train_metadata_ptb, train_features_ptb_raw, by = "specimen")
validation_data_ptb_all <- merge(validation_metadata_ptb, validation_features_ptb_raw, by = "specimen")

# Select last specimen with collect_wk < 32 for each participant in training set
train_data_ptb <- train_data_ptb_all %>%
  filter(collect_wk < 32) %>%
  group_by(participant_id) %>%
  filter(collect_wk == max(collect_wk)) %>%
  ungroup()

# Select last specimen with collect_wk < 32 for each participant in validation set
validation_data_ptb <- validation_data_ptb_all %>%
  filter(collect_wk < 32) %>%
  group_by(participant_id) %>%
  filter(collect_wk == max(collect_wk)) %>%
  ungroup()

# Define PTB outcome (delivery_wk < 37)
train_data_ptb <- train_data_ptb %>%
  mutate(PTB = ifelse(delivery_wk < 37, 1, 0))
validation_data_ptb <- validation_data_ptb %>%
  mutate(PTB = ifelse(delivery_wk < 37, 1, 0))

# Identify common species features
train_species_ptb <- train_features_ptb_raw %>% select(-specimen) %>% colnames()
validation_species_ptb <- validation_features_ptb_raw %>% select(-specimen) %>% colnames()
common_species_ptb <- intersect(train_species_ptb, validation_species_ptb)

# Prepare feature matrices for PTB prediction
train_features_matrix_ptb <- as.matrix(train_data_ptb[, common_species_ptb])
validation_features_matrix_ptb <- as.matrix(validation_data_ptb[, common_species_ptb])
train_outcome_ptb <- train_data_ptb$PTB
validation_outcome_ptb <- validation_data_ptb$PTB

# 2. Model Fitting for PTB (Logistic Regression with LASSO for feature selection)
set.seed(123) # for reproducibility
model_ptb <- cv.glmnet(train_features_matrix_ptb, train_outcome_ptb, family = "binomial", alpha = 1, nfolds = 10) # alpha=1 for LASSO

# 3. Model Evaluation for PTB
predictions_ptb <- predict(model_ptb, newx = validation_features_matrix_ptb, s = "lambda.min", type = "response")
roc_obj_ptb <- roc(validation_outcome_ptb, as.numeric(predictions_ptb))
auc_roc_ptb <- auc(roc_obj_ptb)
print(paste("AUC ROC for PTB prediction:", round(auc_roc_ptb, 3)))

write.csv(data.frame(y_test=validation_outcome_ptb,y_pred=as.numeric(predictions_ptb)),file="./out2/gemini-2A.R.csv")

# Plot ROC curve for PTB
plot(roc_obj_ptb, main = "ROC Curve for PTB Prediction", col = "blue", print.auc = TRUE, auc.polygon = TRUE, grid = c(0.1, 0.2), auc.polygon.col = "#D9F0A3")


# --- EarlyPTB Prediction ---

# 1. Data Loading and Preprocessing for EarlyPTB

# Use the same raw data, just need to re-process based on new criteria

# Select last specimen with collect_wk < 28 for each participant in training set
train_data_earlyptb <- train_data_ptb_all %>% # using _all because reloading is redundant
  filter(collect_wk < 28) %>%
  group_by(participant_id) %>%
  filter(collect_wk == max(collect_wk)) %>%
  ungroup()

# Select last specimen with collect_wk < 28 for each participant in validation set
validation_data_earlyptb <- validation_data_ptb_all %>% # using _all because reloading is redundant
  filter(collect_wk < 28) %>%
  group_by(participant_id) %>%
  filter(collect_wk == max(collect_wk)) %>%
  ungroup()

# Define EarlyPTB outcome (delivery_wk < 32)
train_data_earlyptb <- train_data_earlyptb %>%
  mutate(EarlyPTB = ifelse(delivery_wk < 32, 1, 0))
validation_data_earlyptb <- validation_data_earlyptb %>%
  mutate(EarlyPTB = ifelse(delivery_wk < 32, 1, 0))

# Ensure same common species list as for PTB is used for consistency and as per instructions (common between training and validation for PTB).
# If the instruction meant common species for *each* analysis separately, then re-calculate it.
# For now, assuming common species from PTB analysis should be used.
common_species_earlyptb <- common_species_ptb # Re-use common species from PTB analysis

# Prepare feature matrices for EarlyPTB prediction
train_features_matrix_earlyptb <- as.matrix(train_data_earlyptb[, common_species_earlyptb])
validation_features_matrix_earlyptb <- as.matrix(validation_data_earlyptb[, common_species_earlyptb])
train_outcome_earlyptb <- train_data_earlyptb$EarlyPTB
validation_outcome_earlyptb <- validation_data_earlyptb$EarlyPTB

# Remove rows with NA features, as caused by filtering and species columns not being in the filtered data frames.
# It's safer to ensure no NAs by imputation if necessary, or to filter out rows with NA values *before* model fitting.
# In this case, let's check and remove rows with NA features after subsetting.

na_train_rows_earlyptb <- apply(train_features_matrix_earlyptb, 1, function(x) any(is.na(x)))
train_features_matrix_earlyptb_clean <- train_features_matrix_earlyptb[!na_train_rows_earlyptb, ]
train_outcome_earlyptb_clean <- train_outcome_earlyptb[!na_train_rows_earlyptb]

na_validation_rows_earlyptb <- apply(validation_features_matrix_earlyptb, 1, function(x) any(is.na(x)))
validation_features_matrix_earlyptb_clean <- validation_features_matrix_earlyptb[!na_validation_rows_earlyptb, ]
validation_outcome_earlyptb_clean <- validation_outcome_earlyptb[!na_validation_rows_earlyptb]

# Check if there are any rows left after NA removal
if(nrow(train_features_matrix_earlyptb_clean) == 0 || nrow(validation_features_matrix_earlyptb_clean) == 0) {
  stop("No data left after removing rows with NA features for EarlyPTB. Check data processing steps.")
}


# 2. Model Fitting for EarlyPTB (Logistic Regression with LASSO)
set.seed(123) # for reproducibility
model_earlyptb <- cv.glmnet(train_features_matrix_earlyptb_clean, train_outcome_earlyptb_clean, family = "binomial", alpha = 1, nfolds = 10) # alpha=1 for LASSO

# 3. Model Evaluation for EarlyPTB
predictions_earlyptb <- predict(model_earlyptb, newx = validation_features_matrix_earlyptb_clean, s = "lambda.min", type = "response")
roc_obj_earlyptb <- roc(validation_outcome_earlyptb_clean, as.numeric(predictions_earlyptb))
auc_roc_earlyptb <- auc(roc_obj_earlyptb)
print(paste("AUC ROC for EarlyPTB prediction:", round(auc_roc_earlyptb, 3)))
write.csv(data.frame(y_test=validation_outcome_earlyptb_clean,y_pred=as.numeric(predictions_earlyptb)),file="./out2/gemini-2B.R.csv")

# Plot ROC curve for EarlyPTB
plot(roc_obj_earlyptb, main = "ROC Curve for EarlyPTB Prediction", col = "red", print.auc = TRUE, auc.polygon = TRUE, grid = c(0.1, 0.2), auc.polygon.col = "#FFDAB9")

File: LLMgeneratedCode/gemini-2.py
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score, roc_curve
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler

def analyze_outcome(outcome_type, collect_wk_threshold, delivery_wk_threshold, collect_wk_cutoff):
    """
    Analyzes the prediction of a binary outcome (PTB or EarlyPTB) using microbiome features.

    Args:
        outcome_type (str): "PTB" or "EarlyPTB", the type of preterm birth to predict.
        collect_wk_threshold (int): collect_wk threshold for selecting samples (e.g., 32 for PTB, 28 for EarlyPTB).
        delivery_wk_threshold (int): delivery_wk threshold defining the positive outcome (e.g., 37 for PTB, 32 for EarlyPTB).
        collect_wk_cutoff (int): collect_wk upper bound for filtering initial data.

    Returns:
        None: Prints AUC-ROC score and plots ROC curve for the validation set.
    """

    print(f"\nAnalyzing outcome: {outcome_type}")

    # Load data
    train_feature_df = pd.read_csv('training_species_abundance.csv', index_col=0)
    val_feature_df = pd.read_csv('validation_species_abundance.csv', index_col=0)
    train_metadata_df = pd.read_csv('training_metadata.csv')
    val_metadata_df = pd.read_csv('validation_metadata.csv')

    # Identify common features
    common_features = list(set(train_feature_df.columns) & set(val_feature_df.columns))
    train_feature_df = train_feature_df[common_features]
    val_feature_df = val_feature_df[common_features]

    # --- Process Training Data ---
    train_metadata_filtered = train_metadata_df[train_metadata_df['collect_wk'] < collect_wk_cutoff].copy() # Initial filter before max collect_wk selection

    # Sort by collect_wk to easily pick the last entry
    train_metadata_filtered = train_metadata_filtered.sort_values(by='collect_wk')

    # Get the last entry with collect_wk < collect_wk_threshold for each participant
    train_metadata_last_entries = train_metadata_filtered[train_metadata_filtered['collect_wk'] < collect_wk_threshold].groupby('participant_id').last().reset_index()


    # Merge with feature data
    train_merged_df = pd.merge(train_metadata_last_entries, train_feature_df.reset_index(), on='specimen', how='inner')

    if train_merged_df.empty:
        print(f"No training data available after filtering for {outcome_type}. Check collect_wk thresholds.")
        return


    # Define PTB outcome
    train_merged_df[outcome_type] = (train_merged_df['delivery_wk'] < delivery_wk_threshold).astype(int)


    # --- Process Validation Data ---
    val_metadata_filtered = val_metadata_df[val_metadata_df['collect_wk'] < collect_wk_cutoff].copy() # Initial filter before max collect_wk selection
    val_metadata_filtered = val_metadata_filtered.sort_values(by='collect_wk')
    val_metadata_last_entries = val_metadata_filtered[val_metadata_filtered['collect_wk'] < collect_wk_threshold].groupby('participant_id').last().reset_index()


    # Merge with feature data
    val_merged_df = pd.merge(val_metadata_last_entries, val_feature_df.reset_index(), on='specimen', how='inner')

    if val_merged_df.empty:
        print(f"No validation data available after filtering for {outcome_type}. Check collect_wk thresholds.")
        return

    # Define PTB outcome
    val_merged_df[outcome_type] = (val_merged_df['delivery_wk'] < delivery_wk_threshold).astype(int)


    # Prepare data for modeling
    features = common_features
    X_train = train_merged_df[features]
    y_train = train_merged_df[outcome_type]
    X_val = val_merged_df[features]
    y_val = val_merged_df[outcome_type]

    # Scale features
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_val_scaled = scaler.transform(X_val)

    # Train Logistic Regression model
    model = LogisticRegression(random_state=42, solver='liblinear', class_weight='balanced') # Use class_weight to handle potential imbalance
    model.fit(X_train_scaled, y_train)

    # Make predictions on validation set
    y_pred_proba = model.predict_proba(X_val_scaled)[:, 1]
    auc_roc = roc_auc_score(y_val, y_pred_proba)
    print(f"AUC-ROC on Validation Set for {outcome_type}: {auc_roc:.4f}")

    # Plot ROC curve
    fpr, tpr, thresholds = roc_curve(y_val, y_pred_proba)
    plt.figure(figsize=(8, 6))
    plt.plot(fpr, tpr, label=f'ROC curve (AUC = {auc_roc:.4f})')
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title(f'ROC Curve for {outcome_type} Prediction')
    plt.legend(loc="lower right")
    plt.savefig("./out2/gemini-2.py_"+outcome_type+".pdf", format="pdf")
    df = pd.DataFrame({'y_test': y_val, 'y_prob': y_pred_proba})
    df.to_csv("./out2/gemini-2.py_"+outcome_type+".csv", index=False)


# --- Run analysis for PTB ---
analyze_outcome(outcome_type='PTB', collect_wk_threshold=32, delivery_wk_threshold=37, collect_wk_cutoff=32)

# --- Run analysis for EarlyPTB ---
analyze_outcome(outcome_type='EarlyPTB', collect_wk_threshold=28, delivery_wk_threshold=32, collect_wk_cutoff=28)
